src/python/pants/source/source_root.py

```diff
@@ -26,18 +28,67 @@ class AllSourceRoots(Collection[SourceRoot]):
     pass
 
 
+class SourceRootPatternMatcher:
+    # We perform pattern matching against absolute paths, where "/" represents the repo root.
+    _repo_root = PurePath(os.path.sep)
+
+    def __init__(self, root_patterns: Iterable[str]):
+        self._root_patterns = list(root_patterns)
+
+    def get_patterns(self) -> Tuple[str, ...]:
+        return tuple(self._root_patterns)
+
+    def add_source_root(self, path):
```

Comment:
```suggestion
    def add_source_root(self, path: PurePath) -> None:
```

Reply:
Actually path can be (and usually is) a string.

Reply:
If you're up for it, this would be `Union[str, PurePath]`, then.